DM-17549: Schema compatibility fixes.

jdswinbank opened the pull request.

jdswinbank committed:
Give schema files fully qualified names.

jdswinbank committed:
Move schema into versioned directory.

jdswinbank committed:
Add schema version 1.0

jdswinbank committed:
File example data with associated schemata.

jdswinbank committed:
Add library of Python utils for working with alerts.

jdswinbank committed:
Reimplement round-trip validation using utils.

jdswinbank committed:
Add EUPS table file.

jdswinbank committed:
Add simulation code based on generateFakePackets.py.

jdswinbank committed:
Add routines to store/retrieve alerts to/from file.

jdswinbank committed:
Consolidate many free functions to a Schema class.

jdswinbank committed:
Reimplement alert simulation based on library.

jdswinbank committed:
Add "packet" suffix to module name.

jdswinbank committed:
Add a standalone function to read alerts from file.

Sometimes, we don't have a reader schema, and just want to get whatever the
file has embedded in it.

jdswinbank committed:
Give v1 schema files fully qualified names.

jdswinbank committed:
Add a (trivial) schema registry.

jdswinbank committed:
Resolve schema before hashing.

jdswinbank committed:
Always fully resolve schemata.

Also clear fastavro cache after loading.

See the comments on Schema.from_file for rationale.

ebellm commented on README.rst:
```diff
@@ -0,0 +1,68 @@
+The command will print a brief summary of the size of the data in various formats.
+Thus::
+
+   $ ./bin/validateAvroRoundTrip.py --input-data=./schema/latest/sample_data/alert.json --cutout-template=./examples/stamp-678.fits --cutout--difference=./examples/stamp-679.fits
```
`--cutout-difference` should have only one interior dash

ebellm commented on README.rst:
```diff
@@ -0,0 +1,68 @@
+``bin/validateAvroRoundTrip.py`` demonstrates round-tripping a simple alert through the Avro system.
+Sample data is provided in the ``schema/latest/sample_data/alert.json`` file, or an alternative may be provided on the command line.
+Optionally, the path to binary data files to be included in the packet as “postage stamp” images may be provided.
+If the ``--true`` flag is given, the alert contents are printed to screen for sanity checking.
```
should be `--print`, not `--true`

ebellm commented on README.rst:
```diff
@@ -0,0 +1,68 @@
+-------
+
+The ``lsst.alerts`` Python package provides a suite of routines for working with alerts in the Avro format.
+
```
As we discussed in person, I suggest that we need an `Installation` section here or similar.

In addition to whatever `eups` bits are relevant we should also describe how someone not using the LSST stack can get the commands below working, namely (from the module base directory)

```
export PATH=$PWD/bin:$PATH
export PYTHONPATH=$PWD/python:/$PYTHONPATH
```

ebellm commented on README.rst:
```diff
@@ -0,0 +1,68 @@
+This package provides information about, and utilties for working with, LSST alerts in `Apache Avro`_ format.
+In particular, it includes:
+
+- Alert schemata;
```
Let's use "schemas" in preference to "schemata".  While the latter is more etymologically sound, both are acceptable and the former is both more accessible and more widely used in this context.

https://english.stackexchange.com/questions/77764/plural-form-of-schema

and #dm-naming-things prefers it ;-)

jdswinbank replied:
I bow to the tyranny of the masses. Even when they're wrong. :-)

jdswinbank replied:
Made this change throughout here, lsst-dm/alert_stream and DMTN-093 on this ticket.

ebellm commented on README.rst:
```diff
@@ -0,0 +1,68 @@
+Schemata
+========
+
+Alert schemata are located in the ``schema`` directory.
```
as above, "schemas"

ebellm commented on README.rst:
```diff
@@ -0,0 +1,68 @@
+
+Alert schemata are located in the ``schema`` directory.
+
+Schemata are filed according to their version number, following a ``MAJOR.MINOR`` scheme.
```
as above, "schemas"

ebellm commented on README.rst:
```diff
@@ -0,0 +1,68 @@
+Utility Code
+============
+
+All code is written in Python, and uses the `fastavro`_ library.
```
which may be installed with `pip install fastavro`

jdswinbank replied:
Slightly unhappy with that, because it opens up so many other questions: do folks have permissions to write to the default installation directory, should they actually be using Conda, ...

I'll add a note that “these can both be installed using the standard methods for Python libraries (pip, Conda, etc)”.

ebellm commented on bin/simulateAlerts.py:
```diff
@@ -0,0 +1,43 @@
+def parse_args():
+    parser = argparse.ArgumentParser()
+    # Default value based on LSE-81
+    parser.add_argument('--visits-per-year', type=int, default=1056//10,
```
https://www.lsst.org/scientists/keynumbers would make this `825/10` (no need for float division).  I'm not sure whether that or the LSE-81 sizing estimate used here is preferable.

jdswinbank replied:
Ticketed as [DM-19255](https://jira.lsstcorp.org/browse/DM-19255).

ebellm commented on bin/simulateAlerts.py:
```diff
@@ -0,0 +1,43 @@
+    parser.add_argument('--visits-per-year', type=int, default=1056//10,
+                        help='Number of visits per year.')
+    parser.add_argument('--num-alerts', type=int, default=10,
+                        help='Number of simulated alerts to generate.')
```
Maybe say `Number of simulated alert packets to generate`--as I think there's a bit of ambiguity in what these options do otherwise.

ebellm commented on bin/simulateAlerts.py:
```diff
@@ -0,0 +1,43 @@
+def parse_args():
+    parser = argparse.ArgumentParser()
+    # Default value based on LSE-81
+    parser.add_argument('--visits-per-year', type=int, default=1056//10,
```
It would be useful to me at least to have explicit arguments for `--num-prv-DiaSources`,  `--num-prv-DiaForcedSources`, and `--num-prv-nonDetections` rather than embedding that logic in the `__main__` function below.

jdswinbank replied:
Also included in [DM-19255](https://jira.lsstcorp.org/browse/DM-19255).

ebellm commented on bin/simulateAlerts.py:
```diff
@@ -0,0 +1,43 @@
+    # Default value based on LSE-81
+    parser.add_argument('--visits-per-year', type=int, default=1056//10,
+                        help='Number of visits per year.')
+    parser.add_argument('--num-alerts', type=int, default=10,
```
Let's make the default 1.

jdswinbank replied:
Also included in [DM-19255](https://jira.lsstcorp.org/browse/DM-19255).

ebellm commented on bin/simulateAlerts.py:
```diff
@@ -0,0 +1,43 @@
+    args = parse_args()
+
+    schema = lsst.alert.packet.Schema.from_file()
+    arrayCount = {'prvDiaSources': args.visits_per_year,
```
see above for using explicit arguments here.

jdswinbank replied:
Also included in [DM-19255](https://jira.lsstcorp.org/browse/DM-19255).

ebellm commented on bin/validateAvroRoundTrip.py:
```diff
@@ -0,0 +1,115 @@
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--schema-version', type=str,
+                        help='Schema version to test.', default="latest")
```
I'd explicitly document in the help string that you can specify `"latest"` (default) or `MAJOR.MINOR`.

ebellm commented on bin/validateAvroRoundTrip.py:
```diff
@@ -0,0 +1,115 @@
+"""Demonstrate round-tripping of alert data through Avro serialization.
+"""
+
+# Arguably, this should be a test, rather than an executable.
```
I agree that this is not a script most users will find useful.  Should we just make it a test?

jdswinbank replied:
Ticketed as [DM-19257](https://jira.lsstcorp.org/browse/DM-19257).

ebellm commented on python/lsst/alert/packet/io.py:
```diff
@@ -0,0 +1,42 @@
+        I/O stream from which data will be read.
+    reader_schema : `dict` or `list`, optional
+        A schema describing the contents of the Avro packets. If not provided,
+        the schema used when writing the alert steram will be used.
```
"steram" -> "stream"

ebellm commented on python/lsst/alert/packet/io.py:
```diff
@@ -0,0 +1,42 @@
+    return cutout_dict
+
+def retrieve_alerts(fp, reader_schema=None):
+    """Read alert packets from the given stream.
```
Let's say "from the given I/O stream" to avoid confusion with a Kafka-based stream.

ebellm commented on python/lsst/alert/packet/schema.py:
```diff
@@ -0,0 +1,203 @@
+"""Routines for working with Avro schemata.
```
schemas

ebellm commented on python/lsst/alert/packet/schemaRegistry.py:
```diff
@@ -0,0 +1,114 @@
+"""Provide a lookup table for alert schemata.
```
schemas

ebellm commented on python/lsst/alert/packet/schemaRegistry.py:
```diff
@@ -0,0 +1,114 @@
+__all__ = ["SchemaRegistry"]
+
+class SchemaRegistry(object):
+    """A registry for alert schemata.
```
schemas

ebellm commented on python/lsst/alert/packet/schemaRegistry.py:
```diff
@@ -0,0 +1,114 @@
+        Paramters
+        ---------
+        version : `str`
+            The version by which to look up the schema.
```
I'd provide the expected string format here

ebellm commented on python/lsst/alert/packet/simulate.py:
```diff
@@ -0,0 +1,145 @@
+    This is sufficient because our schemas have unique keys but is
+    not fully general.
+
+    The `argument does not take an `lsst.alert.Schema`, but rather its fully
```
missing tick on argument?

jdswinbank replied:
A missing tick, but also a nonsensical (or, at least, obsolete) comment. Removed entirely.

ebellm commented on python/lsst/alert/packet/schema.py:
```diff
@@ -0,0 +1,203 @@
+        return self.definition == other.definition
+
+    @staticmethod
+    def resolve(schema):
```
(This might be something we could upstream to fastavro--see eg https://github.com/fastavro/fastavro/issues/314)

jdswinbank replied:
Interesting idea —  I hadn't spotted that fastavro issue. Reading it gave me some ideas to improve my implementation, which I've done, and I've added a test.

Figuring out how best to upstream this will take some work (they'll have their own testing standards, and I'll need to figure out how to slot it into their codebase, and I've probably made some naïve mistakes and will have to iterate with them on the implementation. I'll make a ticket to do this, but I won't have time to work on it immediately.

This is [DM-19264](https://jira.lsstcorp.org/browse/DM-19264).

ebellm commented on python/lsst/alert/packet/schema.py:
```diff
@@ -0,0 +1,203 @@
+        records : iterable of `dict`
+            Alert records to be stored.
+        """
+        fastavro.writer(fp, self.definition, records)
```
See discussion on the ticket as to whether this should be `schemaless_writer.`

jdswinbank replied:
Per discussion on [DM-17549](https://jira.lsstcorp.org/browse/DM-17549), we “agree to simply store the schemas with alert packets serialized to disk”. I think that means this code should remain as-is (for now, at least).

ebellm commented on python/lsst/alert/packet/schema.py:
```diff
@@ -0,0 +1,203 @@
+            I/O stream from which data will be read.
+        schema : `list`, optional
+            A schema describing the contents of the Avro packets. If not
+            provided, the schema used when writing the alert steram will be
```
maybe just say "writing the alert" (no "stream")

ebellm commented on python/lsst/alert/packet/io.py:
```diff
@@ -0,0 +1,42 @@
+    records : iterable of `dict`
+        Alert records.
+    """
+    reader = fastavro.reader(fp, reader_schema=reader_schema.definition if reader_schema else None)
```
See discussion on the ticket as to whether this should be `schemaless_reader`; the reader_schema would then be required.  

The Avro format supports schema fingerprinting similar to the confluent wire format, but fastavro does not currently write a schema fingerprint: https://github.com/fastavro/fastavro/issues/201)

jdswinbank replied:
Per discussion on [DM-17549](https://jira.lsstcorp.org/browse/DM-17549), we “agree to simply store the schemas with alert packets serialized to disk”. I think that means this code should remain as-is (for now, at least).

ebellm commented on bin/validateAvroRoundTrip.py:
```diff
@@ -0,0 +1,115 @@
+    parser.add_argument('--schema-version', type=str,
+                        help='Schema version to test.', default="latest")
+    parser.add_argument('--input-data', type=str,
+                        help='Data to serialize.', default=None)
```
The help should make clear that this is JSON-formatted data that matches the supplied schema.

ebellm commented on python/lsst/alert/packet/schemaRegistry.py:
```diff
@@ -0,0 +1,114 @@
+                                     sort_keys=True).encode('utf-8'))
+
+    @classmethod
+    def from_filesystem(cls, root=None, schema_root_file="lsst.alert.avsc"):
```
When I try to run this with no arguments I get an error whose traceback I include below:

```
In [22]: sr = lsst.alert.packet.schemaRegistry.SchemaRegistry.from_filesystem()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-22-9a584e22ca89> in <module>
----> 1 sr = lsst.alert.packet.schemaRegistry.SchemaRegistry.from_filesystem()

~/Eric/ZTF/ZADS/sample-avro-alert/python/lsst/alert/packet/schemaRegistry.py in from_filesystem(cls, root, schema_root_file)
    109         for root, dirs, files in os.walk(root, followlinks=False):
    110             if schema_root_file in files:
--> 111                 schema = Schema.from_file(os.path.join(root, schema_root_file))
    112                 version = ".".join(root.split("/")[-2:])
    113                 registry.register_schema(schema, version)

~/Eric/ZTF/ZADS/sample-avro-alert/python/lsst/alert/packet/schema.py in from_file(cls, filename)
    199             filename = os.path.join(get_schema_root(), "latest", "lsst.alert.avsc")
    200         initial_schema = fastavro.schema.load_schema(filename)
--> 201         resolved_schema = cls.resolve(initial_schema)
    202         fastavro.schema._schema.SCHEMA_DEFS.clear()
    203         return cls(resolved_schema)

~/Eric/ZTF/ZADS/sample-avro-alert/python/lsst/alert/packet/schema.py in resolve(schema)
    162             return output
    163
--> 164         schema_types = {entry['name'] : entry for entry in schema}
    165         schema_root = schema_types.pop('lsst.alert')
    166         return expand_types(schema_root, schema_types)

~/Eric/ZTF/ZADS/sample-avro-alert/python/lsst/alert/packet/schema.py in <dictcomp>(.0)
    162             return output
    163
--> 164         schema_types = {entry['name'] : entry for entry in schema}
    165         schema_root = schema_types.pop('lsst.alert')
    166         return expand_types(schema_root, schema_types)

TypeError: string indices must be integers
> /Users/ebellm/Eric/ZTF/ZADS/sample-avro-alert/python/lsst/alert/packet/schema.py(164)<dictcomp>()
    162             return output
    163
--> 164         schema_types = {entry['name'] : entry for entry in schema}
    165         schema_root = schema_types.pop('lsst.alert')
    166         return expand_types(schema_root, schema_types)
```

jdswinbank replied:
I'm unable to reproduce this. However, I also substantially reworked the code (on 12c86fd2c26492e1238406797efd9f899cd820b6) and added a test (on ec466b06b23389744e3193927851b2fcca14b7cb). Would you mind trying to run the test suite? It should be a matter of simply:
```
$ python -m unittest
```
when you have sample-avro-alert set up.

ebellm replied:
Looks like it was a case of outdated libraries on my side (I was verifying that it worked outside of the LSST stack).

ebellm commented on python/lsst/alert/packet/schema.py:
```diff
@@ -0,0 +1,203 @@
+        -----
+        The interaction with `fastavro` here needs some explanation.
+
+        When `fastavro` loads a schema, it each of the types contained within
```
missing verb after "it"

ebellm reviewed (changes requested):
I'd also suggest a small script in ./bin that would pretty-print an avro packet that had been serialized to disk--this is just a trivial wrapper around `io.retrieve_alerts` but I think would be very useful.

Let's add a few tests!

jdswinbank committed:
Prefer 'schemas' to 'schemata'.

jdswinbank committed:
Improved installation instructions.

jdswinbank committed:
Make it possible to provide validation data on the cmd line.

jdswinbank committed:
Refresh README.

jdswinbank committed:
Simplify schema resolution using fastavro internals.

That is, we refer to fastavro's internal set of schema definitions, rather
than passing around our own.

Also, add a (simplistic) test case for schema resolution.

jdswinbank committed:
Add a test for get_schema_root.

jdswinbank committed:
Refer to schema ID, rather than hash.

It's dangerous to pretend a CRC32 is a hash.

jdswinbank committed:
Test for priming a SchemaRegistry from disk.

jdswinbank committed:
Automatically resolve schemas on construction.

jdswinbank commented:
> I'd also suggest a small script in ./bin that would pretty-print an avro packet that had been serialized to disk--this is just a trivial wrapper around io.retrieve_alerts but I think would be very useful.

Just realised I missed this — I saw the bit about tests, and was so excited about them that I ignored this suggestion. I agree this would be a nice addition. I don't think I want to do it here, but I'll ticket it for an idle moment.
